src/responder.ts

```diff
@@ -51,82 +62,85 @@ enum ResponderEvent {
      */
     constructor() {
         super();
-        this.responseFlow = new ResponseFlow();
     }
 
     // Commodity function to emit events asynchronously
-    protected asyncEmit(...args: any[]) {
-        setImmediate( () => this.emit.call(this, args) );
+    protected asyncEmit(...args: any[]): Promise<boolean> {
+        return new Promise(resolve => resolve(this.emit.apply(this, args)))
     }
-
-    /**
-     * This function tries to submit the transaction that resolves the dispute, protecting the channel.
-     */
-    protected abstract submitStateFunction(): Promise<any>;
-
-    /**
-     * Initiates the response, implementing the strategy of this responder.
-     *
-     * @param responseFlow The ResponseFlow object of this response.
-     */
-    public abstract respond(): Promise<any>;
 }
 
 /**
  * A simple custom Error class to signal that no new block was received while we
  * were waiting for a transaction to be mined. This might likely signal a failure of
  * the provider.
  */
-class NoNewBlockError extends Error {
+export class NoNewBlockError extends Error {
     constructor(message: string) {
         super(message);
         this.name = "NoNewBlockError";
     }
 }
 
+/**
+ * A simple custom Error class to signal that new blocks are being received, but the transaction
+ * is not being mined, suggesting it might be necessary to bump the gas price.
+ */
+export class StuckTransactionError extends Error {
+    constructor(message: string) {
+        super(message);
+        this.name = "StuckTransactionError";
+    }
+}
+
+
 /**
  * A generic abstract responder for the Ethereum blockchain.
+ * It has exclusive control of a wallet, that is, no two instances should share the same wallet.
  * It implements the submitStateFunction, but no strategy.
  */
 export abstract class EthereumResponder extends Responder {
     // TODO-93: the correct gas limit should be provided based on the appointment/integration.
     //          200000 is enough for Kitsune and Raiden (see https://github.com/raiden-network/raiden-contracts/blob/master/raiden_contracts/data/gas.json).
     private static GAS_LIMIT = 200000;
-    /**
-     * @param signer The signer of the wallet associated with this responder. Each responder should have exclusive access to his wallet.
-     * @param appointmentId The id of the Appointment this object is responding to.
-     * @param ethereumResponse The IEthereumResponse containing what needs to be submitted.
-     */
-    constructor(
-        readonly signer: ethers.Signer,
-        public readonly appointmentId: string,
-        public readonly ethereumResponse: IEthereumResponse
-    ) {
+
+    // implementations should query the provider (or a service) to figure out the appropriate gas price
+    protected gasPrice = new ethers.utils.BigNumber(21000000000);
+
+    // the nonce that will be used when submitStateFunction is called
+    protected nonce = null;
```

Comment:
I guess I put null because ethers.js tests if `nonce == null` (in that case it will populate it himself); although `undefined == null` is true, so it doesn't really make a difference here. I didn't want to leave it uninitialized, while keeping the default behaviour of ethers.js.